src/storage/src/source/mysql/snapshot.rs

```diff
@@ -0,0 +1,294 @@
+                .with_consistent_snapshot(true)
+                .with_readonly(true);
+            conn.start_transaction(tx_opts).await?;
+            conn.query_drop("set @@session.time_zone = '+00:00'")
+                .await?;
+
+            // Read the schemas of the tables we are snapshotting
+            // TODO(roshan): Only request the schemas of the tables we are snapshotting
+            let _ = schema_info(&mut conn, vec!["dummyschema"]).await?;
+            // TODO: verify the schema matches the expected schema
+
+            // Send a signal to the leader that we have started our transaction
+            transaction_start_handle.give(&transaction_start_cap_set[0], ()).await;
+            *transaction_start_cap_set = CapabilitySet::new();
+
+            trace!(%id, "timely-{worker_id} started transaction");
+
+            if is_snapshot_leader {
+                // Wait for all workers to start their transactions so we can release the
+                // table locks
+                let mut received = 0;
+                while received < config.worker_count {
+                    match transaction_start_input.next().await {
+                        Some(AsyncEvent::Data(_, _)) => received += 1,
+                        Some(AsyncEvent::Progress(_)) => (),
+                        None => panic!("transaction_start_input closed unexpectedly"),
+                    }
+                }
+
+                trace!(%id, "timely-{worker_id} releasing table locks");
+                if let Some(mut lock_conn) = lock_conn.take() {
+                    lock_conn.query_drop("UNLOCK TABLES").await?;
+                    lock_conn.disconnect().await?;
+                } else {
+                    panic!("lock_conn should have been created for the snapshot leader");
+                }
+                drop(lock_conn);
+            }
+
+            // We have established a snapshot GTID set so we can broadcast the rewind requests
+            for table_qualified_name in reader_snapshot_table_info.keys() {
+                trace!(%id, "timely-{worker_id} producing rewind request for {table_qualified_name}");
+                let req = RewindRequest { table_qualified_name: table_qualified_name.clone(), snapshot_gtid_set: snapshot_gtid_set.clone() };
+                rewinds_handle.give(&rewind_cap_set[0], req).await;
+            }
+            *rewind_cap_set = CapabilitySet::new();
+
+            // Read the snapshot data from the tables
+            let mut final_row = Row::default();
+            let mut temp_storage = Vec::new();
+            for (table_qualified_name, (output_index, table_desc, _casts)) in reader_snapshot_table_info {
+                let query = format!("SELECT * FROM {}", table_qualified_name);
+                trace!(%id, "timely-{worker_id} reading snapshot from table: {table_qualified_name}\n{table_desc:?}");
+                let mut results = conn.query_stream(query).await?;
```

Comment:
note to self: switch to using `conn.exec_stream` https://docs.rs/mysql_async/latest/mysql_async/prelude/trait.Queryable.html#method.exec_stream to use the mysql binary protocol for more efficient decoding https://docs.rs/mysql_async/latest/mysql_async/index.html#mysql-query-protocols